### PR TITLE
Faster crypto by using libssl (#199)

### PR DESCRIPTION
**Pull request by feodoran**

### CODE REVIEW

File: telethon/crypto/aes.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import os
 import pyaes
-
+from . import libssl
```
Comment (Lonami) [inline telethon/crypto/aes.py]: Haven't tested but this will:
```python
raise ImportError('libssl.so not found')
```

If the `'ssl'` library is not found, correct? Shouldn't this rather be:
```python
try:
    from . import libssl
except ImportError:
    libssl = None
```

And then below check `if libssl`?

Comment (feodoran) [inline telethon/crypto/aes.py]: Ah sorry, I mixed something up there. In `libssl.py` I wanted to look like this to:

```
# search and load libssl.so
lib = find_library('ssl')
if not lib:
    libssl = None
else:
    libssl = ctypes.cdll.LoadLibrary(lib)

```

How do I add another commit to this PR?

**Comment by feodoran** [comment]
Somehow there are other older commits and merges in here. Can you somehow clean this up?

**Comment by Lonami** [comment]
Cherry picking should be tbe way to go, although I've never done that. Either that or make a single commit and I'll squash them into one.

**Comment by feodoran** [comment]
What ever works best for you. I am still learning the github workflow ...

File: telethon/crypto/aes.py (hunk 1 above)
Comment (Lonami) [inline telethon/crypto/aes.py]: > How do I add another commit to this PR?

Just push changes to your `crypto` branch and changes should be reflected here.

**Comment by Lonami** [comment]
Thank you :)

**Comment by Lonami** [comment]
@feodoran do you think something similar could be done with the AES-CTR mode?

**Comment by feodoran** [comment]
Yes, I already took a brief look at it. In the header file I found:

```c++
void AES_ctr128_encrypt(const unsigned char *in, unsigned char *out,
                        size_t length, const AES_KEY *key,
                        unsigned char ivec[AES_BLOCK_SIZE],
                        unsigned char ecount_buf[AES_BLOCK_SIZE],
                        unsigned int *num);
```

I don't know yet what the parameters `ecount_buf` and `num` are supposed to be ...

**Comment by Lonami** [comment]
We would have a different problem, there may be several threads trying to access these functions. And every call to the `encrypt` method would need to call `.set_key` right before (which could otherwise be done when creating a new `AES` class, if we were to implement it that way) :/

**Comment by feodoran** [comment]
Sorry, I don't get your point. I think we can just implement it with the class and create the key in `__init__`.

Btw: It would be nice to only import `pyaes` if `libssl` is not found. So one does not need to install `pyaes` if OpenSSL is provided.

**Comment by Lonami** [comment]
My point is, we need to call `AES_set_encrypt_key` when we want to specify a key before calling, for instance, `AES_ige_encrypt`. But when using multiple threads, another thread may call `AES_set_encrypt_key` with a different key, and if the original thread needs the previous key for the other `AES_ige_encrypt` call it won't work. I hope I worded that better now.

> So one does not need to install `pyaes` if OpenSSL is provided.

If it's possible at all to do something like "if OpenSSL is installed, `pyaes` is not a requirement" within `setup.py`, we could surely do that. But first we need to get OpenSSL working for all the `AES`-related operations.

**Comment by feodoran** [comment]
> But when using multiple threads, another thread may call AES_set_encrypt_key with a different key, and if the original thread needs the previous key for the other AES_ige_encrypt call it won't work.

Just create a new independent instance for each thread?

> If it's possible at all to do something like "if OpenSSL is installed, pyaes is not a requirement" within setup.py, we could surely do that. But first we need to get OpenSSL working for all the AES-related operations.

I am not sure whether this will work on all platforms.

**Comment by Lonami** [comment]
> Just create a new independent instance for each thread?

I thought, maybe, we could have a pool with loaded instances that are available.

**Comment by danog** [comment]
A pool with loaded instances indexed by the auth key id.

**Comment by feodoran** [comment]
I don't know when the keys are reused or changed. But that problem would be independent of the implementation of the actual class.

Any ideas about `ecount_buf` and `num`?

**Comment by Lonami** [comment]
Yeah, that problem should be independent of the actual implementation.

About `ecount_buf`, *maybe* you have to initialize it with 0's:
```cpp
/* The input encrypted as though 128bit counter mode is being
 * used.  The extra state information to record how much of the
 * 128bit block we have used is contained in *num, and the
 * encrypted counter is kept in ecount_buf.  Both *num and
 * ecount_buf must be initialised with zeros before the first
 * call to AES_ctr128_encrypt().
 */
```

Taken from [here](https://stackoverflow.com/q/37498912).

**Comment by feodoran** [comment]
Okay, I will try. But what does it count?

Do you have some test/example data for me? Plain, ciper, key, iv.

**Comment by Lonami** [comment]
> But what does it count?

As far as I know it's used as a counter so two identical plaintext blocks don't look the same (since the counter was incremented). It's just a number, given by the IV.

> Do you have some test/example data for me? Plain, ciper, key, iv.

You should be able to use any plain, cipher, key, and IV. Just generate random ones (or simple ones like `b'\x00\x01\x02\x03'`…) and use that data on the known-to-work class linked above.
